### ISSUE TITLE
Surge-VST3: (macOS) linker command failed with exit code 1 (use -v to see invocation) clang

### ISSUE THREAD

**Issue by esaruoho** [kurasu/surge#80]
After proceeding from making a tiny fix ( #79 ) - the next error while trying to get `surge-vst3` built with macOS is:

```
linker command failed with exit code 1 (use -v to see invocation) clang
```

Now, I can't even click it so that xcode would suggest a fix. any ideas?

**Comment by baconpaul** [comment]
Linker errors usually mean you have an undefined symbol. Xcode makes it annoying to find them as far as I can see.

I simulated one in my local area just now

<img width="281" alt="screen shot 2018-12-14 at 6 18 14 pm" src="https://user-images.githubusercontent.com/13387561/50034811-a3201480-ffcc-11e8-9d08-9e55eab087e5.png">

the trick I find to find the symbol is to go to the set of logs and find your failing build

<img width="1277" alt="screen shot 2018-12-14 at 6 18 57 pm" src="https://user-images.githubusercontent.com/13387561/50034842-ccd93b80-ffcc-11e8-8344-b40f24ade361.png">

then when you click on the little expanding hamburger glyph thing you see the symbols you are missing

<img width="633" alt="screen shot 2018-12-14 at 6 19 57 pm" src="https://user-images.githubusercontent.com/13387561/50034852-e1b5cf00-ffcc-11e8-8c02-992f466c5fe2.png">

Usually this occurs because you are including a .h file for a subset of a library but not building or linking the associated .cpp. Especially if you see something like Steinberg::VST::MyBlimBlom it probably means your include path allows you to #include "myblimblom" but you're premake5.lua doesn't include myblimblom.cpp

Hope this helps! Good luck.

**Comment by esaruoho** [comment]
Thanks for the assist! the error i'm seeing is..
<img width="1115" alt="surge_xcworkspace" src="https://user-images.githubusercontent.com/4966687/50045237-d67ba580-0097-11e9-8316-9515d5bce1c6.png">

**Comment by baconpaul** [comment]
Oh yeah

So SurgeVSTEditController isn’t implemented it seems like. But it also isn’t used in the AU as far as I saw. I have an ifdef which comments it out for AU which was in my first get-it-running PR

Is the object even used or is it just referenced?

Seems like an awful lot to implement though. does grep -r show those implemnted anywhere in the source tree? (grep -r “foo” src/ vst3sdk/ is now built into my fingers!)

On Dec 15, 2018, at 11:33 AM, Esa Juhani Ruoho <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks for the assist! the error i'm seeing is..
[surge_xcworkspace]<https://user-images.githubusercontent.com/4966687/50045237-d67ba580-0097-11e9-8316-9515d5bce1c6.png>

**Comment by asimilon** [comment]
I also see these link errors.  Not entirely sure from looking at `git blame` history, so wondering who wrote in the README 

> You can also build a VST which links but honestly we aren't sure if it works. Help from someone who uses VST would be appreciated here!

I'm wondering how this is possible that it built and linked?

**Comment by baconpaul** [comment]
I had written that when it linked! I then got the au working and stopped looking at vst. And apparently now it is back to not linking :)

If it is the surgeditor symbol errors I fixed that by ifdefing out the class which wasn’t used. Later tonight I will take a peek and see what’s in head.

These sorts of problems will pop up more and more the longer our merge backlog is! I really hope we can get master back close to the codebase which runs the au soon.

**Comment by baconpaul** [comment]
Right this diff 

https://github.com/kzantow/surge/commit/f20830627f7b0e6546b771ab2f087a537c9673ea

Which made it work. That if 0 went to if !target audio unit, though, which is why au works and vst no longer does

So to make vst work also exclude that class if it is truly not needed or finish those methods. With it excluded I could compile and link but not run a vst

**Comment by baconpaul** [comment]
Oh and that is my diff. Just I’m looking st it on kzantows big merge branch which is why his name shows up.

**Comment by asimilon** [comment]
Slightly unrelated, but on the VST2 project, with some excluding of AU specific files from `premake5.lua` I've got it successfully building and linking, however, it's not being recognised.  Perhaps it's going to be similar issue for the VST3, will investigate.

**Comment by esaruoho** [comment]
@asimilon query -- what is not recognizing Surge.vst? Have you checked https://github.com/tracktion/pluginval to see what it says on the VST?

**Comment by baconpaul** [comment]
great! Happy to answer any questions about the au or look at any diffs if that helps

**Comment by baconpaul** [comment]
OK I figured out the premake thing that was breaking vat and how to change it so that AU still worked and have a diff coming which will allow VST3 and AU to link. Stay tuned. I'll push it to the big PR.

**Comment by baconpaul** [comment]
https://github.com/kurasu/surge/pull/92/commits/ce3b57d913845e69d3148a1d66e176d52cbeb8ce

OK there's the diff. @asimilon if you want to use this you need to fetch and merge @kzantow branch "cumulative-fixes-and-zoomable-ui"

I have not included your build.osx changes in that. Honestly it was too tricky to pull the commit from a fourth remote! 

But with those changes, VST3 and AU both link if you start from a clean area and rerun premake

**Comment by baconpaul** [comment]
Oh and by being in cumulative-fixes-and-zoomable-ui it participates in the big PR which gets the AU running also.

I don't want to add a million things to that PR. If you actually can get the VST validating from there maybe a separate one makes sense but to do that one you will probably want to start from the cumulative-fixes branch.

**Comment by baconpaul** [comment]
Can we close this issue @esaruoho - It's now resolved after the #92 merge.

**Comment by esaruoho** [comment]
Closing ticket :)
